### ISSUE TITLE
Average response length is decreasing with PPO Trainer generate

### ISSUE THREAD

**Issue by abaheti95** [huggingface/trl#986]
Hello,
I'm using PPO trainer for a Helpful and Harmless assistant dataset with an existing reward model. I set the `generation_kwargs` similar to the example script (except `"max_new_tokens"=128`). https://github.com/huggingface/trl/blob/9e9f024399b76842ece3552884bbc4f304fd4153/examples/scripts/ppo.py#L180-L187

However, no matter the starting checkpoint for PPO, it consistently decreases the average response length as training progresses. 
I'm using a rollout batch size of 128, internal mini-batch size of 16, and 4 internal ppo_epochs. I am also using PPOTrainer's generation function as follows `ppo_trainer.generate(ppo_prefix_tensors,return_prompt=False,**generation_kwargs)`
Here, is a log of average response tokens from two different runs with different starting checkpoints:

```json
"mean|std"
{"step": 1, "ppo_response_length": "52.5469|37.7781"}
{"step": 2, "ppo_response_length": "58.9375|39.3257"}
{"step": 3, "ppo_response_length": "49.6562|39.3231"}
{"step": 4, "ppo_response_length": "52.4766|40.9758"}
{"step": 5, "ppo_response_length": "48.2422|38.0722"}
{"step": 6, "ppo_response_length": "58.8906|40.9970"}
{"step": 7, "ppo_response_length": "50.9531|37.7485"}
{"step": 8, "ppo_response_length": "43.5703|34.9264"}
{"step": 9, "ppo_response_length": "44.6328|35.6909"}
{"step": 10, "ppo_response_length": "39.0781|32.4886"}
{"step": 11, "ppo_response_length": "35.0703|29.9278"}
{"step": 12, "ppo_response_length": "37.5391|32.8386"}
{"step": 13, "ppo_response_length": "31.3594|30.5020"}
{"step": 14, "ppo_response_length": "31.3281|29.7655"}
{"step": 15, "ppo_response_length": "24.1094|22.6599"}
{"step": 16, "ppo_response_length": "26.2266|26.3727"}
```

```json
"mean|std"
{"step": 1, "ppo_response_length": "93.8828|42.4514"}
{"step": 2, "ppo_response_length": "86.9922|44.3400"}
{"step": 3, "ppo_response_length": "83.2266|44.6506"}
{"step": 4, "ppo_response_length": "81.2344|46.9773"}
{"step": 5, "ppo_response_length": "74.9219|45.9622"}
{"step": 6, "ppo_response_length": "76.8906|45.2468"}
{"step": 7, "ppo_response_length": "67.5703|46.7145"}
{"step": 8, "ppo_response_length": "55.3906|41.3116"}
{"step": 9, "ppo_response_length": "60.1172|46.0400"}
{"step": 10, "ppo_response_length": "63.2656|43.7828"}
{"step": 11, "ppo_response_length": "55.5938|41.1258"}
{"step": 12, "ppo_response_length": "59.0547|43.8095"}
{"step": 13, "ppo_response_length": "37.6328|33.5251"}
{"step": 14, "ppo_response_length": "36.3359|31.5746"}
{"step": 15, "ppo_response_length": "35.1641|33.2676"}
{"step": 16, "ppo_response_length": "32.1719|30.8153"}
```

With prolonged training steps the average generation length reduces to just 20 or fewer tokens on average.
Can anyone suggest any solutions to fix this issue?

**Comment by aliwalker** [comment]
I've also been struggling with this length decreasing issue. I've tried to adjust some hyperparams but still have not solved it yet. My guess now is that the scale of reward score(obtained from the reward model) and KL is causing the problem. Since the per token KL penalty is propagated from EOS, reducing the length of response can also reduce the cumulative KL(when calculating the return).

**Comment by lvwerra** [comment]
It could also be related to the reward model that rewards shorter responses higher. You could add an additional reward term directing the reward towards the desired length.

**Comment by prd-tuong-nguyen** [comment]
@abaheti95 hey bro, do you resolve this problem?

**Comment by abaheti95** [comment]
No, I couldn't. Still getting decreasing length consistently. I tried a bigger reward model from 1.4B -> 7B and experimented with a smaller learning rate. But nothing worked. What is your experience @prd-tuong-nguyen ?

**Comment by prd-tuong-nguyen** [comment]
@abaheti95 The same, my response length also decreased by the time. Currently I not find any solution to resolve this problem :((( Due to this problem, it make some unnatural response :((
![image](https://github.com/huggingface/trl/assets/145306938/e0e4a9f7-0ab4-42e0-8bea-c52713e17f3b)

**Comment by Huan80805** [comment]
@prd-tuong-nguyen @abaheti95 Hi guys, did you resolve this? I bumped into the same question...
I am wondering if a length penalty can help?

**Comment by abaheti95** [comment]
Length penalty may help, I haven't experimented with it, however. 

The problem could also be due to a poor reward model or a bad base policy. In my case, I noticed the ppo model optimized safety and subsequently got shorter responses (which also received high reward).
